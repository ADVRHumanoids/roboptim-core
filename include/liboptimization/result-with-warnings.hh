// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the ResultWithWarnings class.
 */

#ifndef OPTIMIZATION_RESULT_WITH_WARNINGS_HH
# define OPTIMIZATION_RESULT_WITH_WARNINGS_HH
# include <vector>

# include <liboptimization/fwd.hh>
# include <liboptimization/result.hh>
# include <liboptimization/solver-warning.hh>

namespace optimization
{
  /// Retrieved when a result has been found but some problems
  /// happened during the computation.
  class ResultWithWarnings : public Result
  {
  public:
    explicit ResultWithWarnings (const Function::size_type,
                                 const Function::size_type = 1)
      throw ();
    ~ResultWithWarnings () throw ();

    std::vector<SolverWarning> warnings;

    virtual std::ostream& print (std::ostream& o) const throw ();
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_RESULT_WITH_WARNINGS_HH
