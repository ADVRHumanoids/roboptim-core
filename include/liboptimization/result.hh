// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Result class.
 */

#ifndef OPTIMIZATION_RESULT_HH
# define OPTIMIZATION_RESULT_HH
# include <iostream>
# include <vector>

# include <liboptimization/function.hh>

namespace optimization
{
  /// \brief Result class, used to communicate the result.
  class Result
  {
  public:
    explicit Result (const Function::size_type n_) throw ();
    explicit Result (const Result&) throw ();
    Result* operator= (const Result& r) throw ();
    virtual ~Result () throw ();

    virtual std::ostream& print (std::ostream& o) const throw ();

    const Function::size_type n;
    Function::vector_t x;
    Function::value_type value;
    Function::vector_t lambda;
  };

  /// Retrieved when a result has been found but some problems
  /// happened during the computation.
  class ResultWithWarnings : public Result
  {
  public:
    ~ResultWithWarnings () throw ();

    std::vector<SolverWarning*> warnings;

    virtual std::ostream& print (std::ostream& o) const throw ();
  };

  std::ostream& operator<< (std::ostream& o, const Result& r);

} // end of namespace optimization

#endif //! OPTIMIZATION_RESULT_HH
