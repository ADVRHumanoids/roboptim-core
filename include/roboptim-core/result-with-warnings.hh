// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the ResultWithWarnings class.
 */

#ifndef ROBOPTIM_CORE_RESULT_WITH_WARNINGS_HH
# define ROBOPTIM_CORE_RESULT_WITH_WARNINGS_HH
# include <vector>

# include <roboptim-core/fwd.hh>
# include <roboptim-core/result.hh>
# include <roboptim-core/solver-warning.hh>

namespace roboptim
{
  /// Retrieved when a result has been found but some problems
  /// happened during the computation.
  class ResultWithWarnings : public Result
  {
  public:
    explicit ResultWithWarnings (const Function::size_type,
                                 const Function::size_type = 1)
      throw ();
    ~ResultWithWarnings () throw ();

    std::vector<SolverWarning> warnings;

    virtual std::ostream& print (std::ostream& o) const throw ();
  };

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_RESULT_WITH_WARNINGS_HH
