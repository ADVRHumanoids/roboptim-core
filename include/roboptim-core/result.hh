// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Result class.
 */

#ifndef ROBOPTIM_CORE_RESULT_HH
# define ROBOPTIM_CORE_RESULT_HH
# include <iostream>
# include <vector>

# include <roboptim-core/function.hh>

namespace roboptim
{
  /// \brief Result class, used to communicate the result.
  class Result
  {
  public:
    explicit Result (const Function::size_type,
                     const Function::size_type = 1) throw ();
    explicit Result (const Result&) throw ();
    Result* operator= (const Result& r) throw ();
    virtual ~Result () throw ();

    virtual std::ostream& print (std::ostream& o) const throw ();

    const Function::size_type n;
    const Function::size_type m;
    Function::vector_t x;
    Function::vector_t value;
    Function::vector_t lambda;
  };

  std::ostream& operator<< (std::ostream& o, const Result& r);
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_RESULT_HH
