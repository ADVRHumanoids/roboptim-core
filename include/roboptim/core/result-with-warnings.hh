// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_RESULT_WITH_WARNINGS_HH
# define ROBOPTIM_CORE_RESULT_WITH_WARNINGS_HH

# include <roboptim/core/result.hh>
# include <roboptim/core/portability.hh>

namespace roboptim
{
  /// \addtogroup roboptim_problem
  /// @{

  /// \brief Warnings have been merged to Result to simplify the API.
  typedef Result ResultWithWarnings ROBOPTIM_CORE_DEPRECATED;

  /// @}

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_RESULT_WITH_WARNINGS_HH
