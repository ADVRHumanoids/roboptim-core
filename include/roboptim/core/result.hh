// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Result class.
 */

#ifndef ROBOPTIM_CORE_RESULT_HH
# define ROBOPTIM_CORE_RESULT_HH
# include <iostream>
# include <vector>

# include <roboptim/core/function.hh>

namespace roboptim
{
  /// \brief Result class, used to communicate the result.
  class Result
  {
  public:
    typedef Function::size_type size_type;
    typedef Function::vector_t vector_t;

    explicit Result (const size_type,
                     const size_type = 1) throw ();
    virtual ~Result () throw ();

    virtual std::ostream& print (std::ostream& o) const throw ();

    size_type inputSize;
    size_type outputSize;
    vector_t x;
    vector_t value;
    vector_t lambda;
  };

  std::ostream& operator<< (std::ostream& o, const Result& r);
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_RESULT_HH
