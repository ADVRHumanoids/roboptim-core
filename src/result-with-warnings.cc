// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Result class.
 */

#include <iostream>
#include <vector>

#include <boost/numeric/ublas/io.hpp>

#include <roboptim/core/function.hh>
#include <roboptim/core/indent.hh>
#include <roboptim/core/result-with-warnings.hh>
#include <roboptim/core/solver-warning.hh>
#include <roboptim/core/util.hh>

namespace roboptim
{
  ResultWithWarnings::ResultWithWarnings (const Function::size_type n,
                                          const Function::size_type m)
    throw ()
    : Result (n, m)
  {
  }

  std::ostream&
  ResultWithWarnings::print (std::ostream& o) const throw ()
  {
    Result::print (o);
    return o << incindent << iendl << "Warnings: " << warnings << decindent;
  }

  ResultWithWarnings::~ResultWithWarnings () throw ()
  {
  }

} // end of namespace roboptim
