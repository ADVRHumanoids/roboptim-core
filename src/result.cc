// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Result class.
 */

#include <iostream>
#include <vector>

#include <boost/numeric/ublas/io.hpp>

#include <liboptimization/function.hh>
#include <liboptimization/indent.hh>
#include <liboptimization/result.hh>
#include <liboptimization/util.hh>

namespace optimization
{
  Result::Result (const Function::size_type n_) throw ()
    : n (n_),
      x (n),
      value (),
      lambda (n)
  {
  }

  Result::Result (const Result& r) throw ()
    : n (r.n),
      x (r.x),
      value (r.value),
      lambda (r.lambda)
  {
  }

  Result*
  Result::operator= (const Result& r) throw ()
  {
    delete this;
    return new Result (r);
  }

  Result::~Result () throw ()
  {
  }

  std::ostream&
  Result::print (std::ostream& o) const throw ()
  {
    o << "Result: " << incindent << iendl
      << "Dimension: " << n << iendl
      << "X: " << x << iendl
      << "Value: " << value << iendl
      << "Lambda: " << lambda
      << decindent;
    return o;
  }

  std::ostream& operator<< (std::ostream& o, const Result& r)
  {
    return r.print (o);
  }

} // end of namespace optimization
