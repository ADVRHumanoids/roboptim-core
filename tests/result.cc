// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/result.hh>
#include <roboptim/core/result-with-warnings.hh>

using namespace roboptim;

// Define a simple function.
struct F : public Function
{
  F () : Function (1, 1, "x")
  {}

  void impl_compute (result_ref res,
                     const_argument_ref x) const
  {
    res (0) = x[0];
  }

  // No gradient, hessian.
};

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (result)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("result");

  // Instantiate some results.
  Result result (3, 8);
  F f;
  F::size_type c_size = 4;
  result.constraints.resize (c_size);
  result.constraints.setZero ();
  result.lambda.resize (c_size);
  result.lambda.setZero ();
  result.constraint_violation = 0.;

  assert (result.x.size () == 3);
  assert (result.value.size () == 8);
  assert (result.lambda.size () == c_size);

  Result result2 (result);
  Result result3 = result;

  (*output) << result << std::endl;

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
