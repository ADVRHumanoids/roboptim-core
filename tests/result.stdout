Result: 
  Size (input, output): 3, 8
  X: [3](0,0,0)
  Value: [8](0,0,0,0,0,0,0,0)
  Lambda: [0]()
